### PR TITLE
Fix incorrect size determination in GetSizeForThemeRoom

### LINKED ISSUE #4957: Incorrect room size for Catacombs
**Important information**
7bb401d9e8b0cca0f8ace40243fc9560eeb2282d changed how rooms in the catacombs are generated :sweat_smile: 

**To Reproduce**
Load [single_3.zip](https://github.com/diasurgical/devilutionX/files/9080298/single_3.zip)
See misplaced door tile:
image

**Additional context**
A test case for this issue https://github.com/diasurgical/devilutionX/pull/4958

### PR DESCRIPTION
**Pull request by ephphatha**
This was a corner case where a boundary was determined by the edge of the search region. Not 100% sure it's correct in all cases, aiming to add some tests to try cover every possible case.

Fixes #4957 and doesn't break existing tests however the cases in the last loop could be expressed better for sure.

### CODE REVIEW

File: Source/levels/gendung.cpp [hunk 1]
```diff
@@ -123,27 +123,26 @@ std::optional<Size> GetSizeForThemeRoom(int floor, Point origin, int minSize, in
 		}
 	}
 
-	// area is at least as high as necessary. If we wanted to find the largest rectangular area we should keep going
-	// and start checking for the largest total area we can find (could be the tallest region that maintains current
-	// width, or maybe a narrower but taller region has a larger area). Instead to match vanilla Diablo logic we
-	// trigger a break as soon as the width shrinks again.
-	for (int yOffset = minSize; yOffset < maxHeight; yOffset++) {
-		for (int xOffset = 0; xOffset < maxWidth; xOffset++) {
+	// Worth out the tallest area we could potentially fill
```
Comment (AJenbo) [inline Source/levels/gendung.cpp:126]: ?
```suggestion
	// Work out the tallest area we could potentially fill
```
